Foundation/NSDecimal.swift

```diff
@@ -1,155 +1,1837 @@
+}
+
+fileprivate func integerMultiplyByPowerOf10(_ result: inout WideDecimal, _ left: WideDecimal, _ p: Int) -> NSDecimalNumber.CalculationError {
+    var power = p
+    if power == 0 {
+        result = left
+        return .noError
+    }
+    let isNegative = power < 0
+    if isNegative {
+        power = -power
+    }
+    result = left
+
+    let maxpow10 = pow10.count - 1
+    var error:NSDecimalNumber.CalculationError = .noError
+
+    while power > maxpow10 {
+        var big = WideDecimal()
+
+        power -= maxpow10
+        let p10 = WideDecimal(pow10[maxpow10])
+
+        if !isNegative {
+            error = integerMultiply(&big,result,p10)
+        } else {
+            error = integerDivide(&big,result,p10)
+        }
+
+        if error != .noError && error != .lossOfPrecision {
+            return error;
+        }
+
+        for i in 0..<Decimal.maxSize {
+            result[i] = big[i]
+        }
+
+        result._length = big._length
+    }
+
+    var big = WideDecimal()
+    // Handle the rest of the power (<= maxpow10)
+    let p10 = WideDecimal(pow10[Int(power)])
+
+    if !isNegative {
+        error = integerMultiply(&big, result, p10)
+    } else {
+        error = integerDivide(&big, result, p10)
+    }
+
+    for i in 0..<big._length {
+        result[i] = big[i]
+    }
+
+    result._length = big._length
+
+    return error;
+}
+
+public func NSDecimalRound(_ result: UnsafeMutablePointer<Decimal>, _ number: UnsafePointer<Decimal>, _ scale: Int, _ roundingMode: NSDecimalNumber.RoundingMode) {
+    NSDecimalCopy(result,number) // this is unnecessary if they are the same address, but we can't test that here
+    result.pointee.round(scale: scale,roundingMode: roundingMode)
+}
 // Rounds num to the given scale using the given mode.
 // result may be a pointer to same space as num.
 // scale indicates number of significant digits after the decimal point
 
-public func NSDecimalNormalize(_ number1: UnsafeMutablePointer<Decimal>, _ number2: UnsafeMutablePointer<Decimal>, _ roundingMode: NSDecimalNumber.RoundingMode) -> NSDecimalNumber.CalculationError { NSUnimplemented() }
+public func NSDecimalNormalize(_ a: UnsafeMutablePointer<Decimal>, _ b: UnsafeMutablePointer<Decimal>, _ roundingMode: NSDecimalNumber.RoundingMode) -> NSDecimalNumber.CalculationError {
+    var diffexp = a.pointee.__exponent - b.pointee.__exponent
+    var result = WideDecimal()
+
+    //
+    // If the two numbers share the same exponents,
+    // the normalisation is already done
+    //
+    if diffexp == 0 {
+        return .noError
+    }
+
+    //
+    // Put the smallest of the two in aa
+    //
+    var aa: UnsafeMutablePointer<Decimal>
+    var bb: UnsafeMutablePointer<Decimal>
+
+    if diffexp < 0 {
+        aa = b
+        bb = a
+        diffexp = -diffexp
+    } else {
+        aa = a
+        bb = b
+    }
+
+    //
+    // Build a backup for aa
+    //
+    var backup = Decimal()
+
+    NSDecimalCopy(&backup,aa)
```

Comment:
NB this seems redundant at the moment but will be used in the NSUnimplemented below